### ISSUE TITLE
Don't repeat tracks in shuffle mode

### ISSUE THREAD

**Issue by eayus** [nukeop/nuclear#1125]
**Platform:**
Void Linux

**Nuclear version:**
0.6.17

**Description of the issue:**
In shuffle mode, it would be nice if shuffle did not repeat any tracks until all of the other tracks in the queue have been exhausted. I think tracks should only be repeated once all the tracks in the queue have been played once.

**Comment by nukeop** [comment]
Currently shuffle has no memory, it just picks a random track every time. Does it work like this in other players?

**Comment by eayus** [comment]
I believe spotify shuffles all the tracks in the queue at the start, and then plays through them 1 by 1 (in the new shuffled order). I would prefer this behaviour, since sometimes with nuclear I'll get the same track playing twice in a row.

**Comment by terryyLi** [comment]
Hi, I wonder would you like it to be when click next button in shuffle mode, just stop and delete the current song from queue and randomly choose a next song to play?

**Comment by nukeop** [comment]
I don't think it should delete anything, just keep track of what's already been played.

**Comment by socherbyc** [comment]
In NewPipe, all songs in the queue are shuffled and played one by one in order. This way, I can even make adjustments to the order (e.g., move similar songs away)

**Comment by BWPanda** [comment]
I think the workflow could be something like this (starting from an empty queue)...

If Shuffle is disabled:
- Clicking Play adds all songs to queue in normal order
- If Shuffle is then enabled, queue is shuffled

If Shuffle is enabled:
- Clicking Play adds all songs to queue in random order
- If Shuffle is then disabled, nothing happens...?

**Comment by nukeop** [comment]
Ok, I think I like this idea. It makes it explicit how the queue is going to be played. I'm not sure though if it should only shuffle the part that is yet to be played, leaving the previously played songs untouched? Or should it shuffle the playlist, then move to the first item in the queue?

**Comment by f03n1x** [comment]
I was kinda wondering about this, I have a specific playlist I listen to while working and notice that on occasion it will indeed play the same song back to back sometimes. I think one way to at least resolve this in a specific circumstance (when playlist isn't ever changing) is using something like the Fisher-Yates shuffle, it's something I used in one of the games I was working on a few years ago, this was so that we'd never have enemy mobs spawn on the same spot, it's also used for something like shuffling a deck of cards. Here is a link to something I found awhile go that help me lead down the path of using it: https://www.dotnetperls.com/shuffle I could maybe look into implementing this myself just because I'd love to at least try my hand at helping a project (especially one I use on a daily basis!).

**Comment by nukeop** [comment]
This issue is from an older version of Nuclear that has been replaced with a complete rewrite. None of the old issues apply to the new codebase.

The old code is preserved on the [legacy/electron](https://github.com/nukeop/nuclear/tree/legacy/electron) branch. If you have feedback about the new version, head to [Discussions](https://github.com/nukeop/nuclear/discussions).
